aten/src/ATen/native/native_functions.yaml

```diff
@@ -3909,6 +3909,9 @@
   variants: function, method
   device_check: NoCheck
   device_guard: False
+  dispatch:
+    CompositeImplicitAutograd: reshape
+    NestedTensorCPU, NestedTensorCUDA: reshape_nested
```

Comment:
I kinda doubt this is related to the inference_mode() issue, but this looks a bit suspicious. Do we want `NestedTensor::reshape` to be able to work with autograd? If we do, some quick comments:

The way that normal `reshape()` works today is that it sometimes returns a of the input, and sometimes doesn't depending on the arguments to reshape. We can't actually support an autograd formula for that directly, so what happens instead of that `reshape()` decomposes into a view op or a clone, and autograd handles those individual ops instead. Does the nested tensor implementation of `reshape()` also "sometimes return a view"?

Reply:
@bdhirsh Yes that's correct, AFAICT the NT implementation of reshape uses the same underlying logic to determine when to return a view and when to return a copy. Do you see it being possible to define a NT-specific reshape_backward that deals with this properly?

Note that we need a NT-specific version of reshape because the semantics of -1s are slightly different due to the possibility for ragged dims.

Reply:
Nested tensor implementation can return a view but I guess that's not the "view" autograd engine uses:
* I heard that autograd would handle "view" differently as "fast path"
* I also heard that this "fast path" is disabled for nested tensor: it would record nested tensor view just as another function
* So based on sharing memory it can return a view, but based on "fast path" it's not an autograd view I think

Could that trigger some edge case?

Reply:
> @bdhirsh Yes that's correct, AFAICT the NT implementation of reshape uses the same underlying logic to determine when to return a view and when to return a copy. Do you see it being possible to define a NT-specific reshape_backward that deals with this properly?
> 
> Note that we need a NT-specific version of reshape because the semantics of -1s are slightly different due to the possibility for ragged dims.

Yes the way I implemented reshape_backward is just like Driss's way for linear. Although I can also do the ugly "is_nested()" version to maintain compositeimplicit-ness

Reply:
> Yes the way I implemented reshape_backward is just like Driss's way for linear.

This is good, but I'm curious how this backward deals with the forward sometimes returning a view and sometimes returning a copy?

Reply:
> > Yes the way I implemented reshape_backward is just like Driss's way for linear.
> 
> This is good, but I'm curious how this backward deals with the forward sometimes returning a view and sometimes returning a copy?

Basically no special treatment: simply reshape grad as input